Feature/message signing

ivolz opened the pull request.

johannesschneiders committed:
WIP: Changes for message signing in the interceptor

Commit:
Add flag to use mutex if message signing is enabled only

Commit:
Revert "Add flag to use mutex if message signing is enabled only"

This reverts commit 1c9b3e2bacbe610eabf9738524b017a92583242b.

Commit:
Message signing should check for Approov-Token header inclusion; disable changing header name if message signing has been enabled

Commit:
Implement message signing changes to apply to different domains and map to configurations

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovMessageSigningConfig.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.ArrayList;
+
+/* message signing configuration
+* This class is used to configure the message signing feature. The message signature can be computed based on the
```
This comment needs to be updated to reflect the changes to the class. It should also explain what the message signing behaviour is, which bits are included in the message, the order and any transformation that is performed (b64 encode the body).

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovMessageSigningConfig.java:
```diff
@@ -0,0 +1,87 @@
+* You can have multiple configurations for different domains and a default '*' configuration that will be used if no
+* specific configuration is found for a domain.
+*/
+public class ApproovMessageSigningConfig {
```
We should add a new interface as follows:

interface MessageSigningConfig {
  public String getTargetHeader();
  public String generateSigningMessage(Request request, String approovTokenHeader);
  public String generateTargetHeaderValue(String messageSignature, String approovTokenHeader);
}

then, this class should instead be called DefaultMessageSigningConfig and it should implement MessageSigningConfig.

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovMessageSigningConfig.java:
```diff
@@ -0,0 +1,87 @@
+        return targetHeader;
+    }
+    /* Get signed headers */
+    public ArrayList<String> getSignedHeaders() {
```
I don't think we need the getter, I can't think of any reason that would be necessary.

ivolz replied:
Debugging purposes perhaps only?

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovMessageSigningConfig.java:
```diff
@@ -0,0 +1,87 @@
+        return signedHeaders;
+    }
+    /* Add a header to the list of signed headers: NOTE the sequence of headers DOES matter */
+    public void addSignedHeader(String header) {
```
It makes sense to make addSignedHeader chainable to simplify the configuration code. So it would be:
public DefaultMessageSigningConfig addSignedHeader(String header); (returning 'this' at the end of the method)

It would make configuration work like this:
aConfig = new DefaultMessageSigningConfig("header").
    addSignedHeader("header1").
    addSignedHeader("header2");

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovMessageSigningConfig.java:
```diff
@@ -0,0 +1,87 @@
+    // the list of headers to include in the message to be signed, in the order they should be added
+    protected ArrayList<String> signedHeaders;
+    // true if the message body should also be signed
+    protected Boolean signBody;
```
After chatting with Richard - remove the following: signBody, signApproovToken, signURL, and signMethod

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovService.java:
```diff
@@ -862,6 +903,69 @@ else if (approovResults.getStatus() != Approov.TokenFetchStatus.UNKNOWN_KEY)
+                // If no default configuration is found, do not sign the message
+                return chain.proceed(request);
+            }
+            Log.d(TAG, "Signing message with configuration for domain: "  + domain + ":  " + (messageSigningConfig.getSignApproovToken() ? ", " + approovTokenHeader +  " **** " : "") + " headers " + messageSigningConfig.getSignedHeaders() +
```
This code should all move into the "public String generateMessage(request Request, String approovTokenHeader);" in the DefaultMessageSigningConfig implementation.

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovService.java:
```diff
@@ -862,6 +903,69 @@ else if (approovResults.getStatus() != Approov.TokenFetchStatus.UNKNOWN_KEY)
+            // 3. add the Approov token to the message, if required (TODO: Note we add the approov token and the prefix!?????)
+            if (messageSigningConfig.signApproovToken) {
+                if (approovResults.getToken() != null) {
+                    message.append(approovTokenPrefix + approovResults.getToken());
```
You should read the approov header from the request instead of recreating the contents here - it is just a safer approach as the same process will occur on the server side -once you migrate the code, into the DefaultRequestSigningConfig you will have to do that anyway.

jexh reviewed (commented):
@ivolz I think my comments indicate the direction we should go but it is worth checking with @richardmtaylor tomorrow to make sure we are all on the same page.

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovService.java:
```diff
@@ -862,6 +903,69 @@ else if (approovResults.getStatus() != Approov.TokenFetchStatus.UNKNOWN_KEY)
+                if (body != null) {
+                    Buffer buffer = new Buffer();
+                    body.writeTo(buffer);
+                    message.append(buffer.readUtf8());
```
We need to b64 encode the body to make sure there are no control characters in the string when it is passed to the approov SDK signing method.

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovService.java:
```diff
@@ -82,6 +83,9 @@ public class ApproovService {
     private static Map<String, Pattern> exclusionURLRegexs = null;
 
+    // message signing configurations mapped to a domain
+    private static Map<String,ApproovMessageSigningConfig> messageSigningConfigs = null;
```
Richard convinced me to remove the map - just have a single config. (@richardmtaylor : I'm now a bit concerned about the different headers that will be used for different endpoints / apis - if we include all of these header names in the message signing target header for all requests are we not potentially leaking information about other APIs that are in use in the app to 3rd parties that shouldn't have this information? - having a config per domain would hopefully avoid this.)

richardmtaylor replied:
It's a good point, but I'm not sure it is worth the complexity to hide this. We will only send it on domains that have Approov tokens added anyway. We shouldn't be signing domains that are not Approov protected, since we rely on an Approov token to prevent replay attacks anyway. And all we are leaking is the names of headers, so don't think it is that big of a deal.

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovService.java:
```diff
@@ -166,8 +195,12 @@ public static synchronized void setDevKey(String devKey) throws ApproovException
-    public static synchronized void setApproovHeader(String header, String prefix) {
+    public static synchronized void setApproovHeader(String header, String prefix) throws ApproovException {
         Log.d(TAG, "setApproovHeader " + header + ", " + prefix);
+        // We only allow this to proceed if message signing is not enabled since changing the header after configuring the message signing is not allowed
```
No need to throw here anymore. The active approov header name is passed to the message signing config every time it processes a request.

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovService.java:
```diff
@@ -862,6 +903,69 @@ else if (approovResults.getStatus() != Approov.TokenFetchStatus.UNKNOWN_KEY)
+            // compute the signature and add it to the request (passing on any exceptions that may occur)
+            if (messageSigningConfig.targetHeader != null && !messageSigningConfig.targetHeader.isEmpty()) {
+                String signature = ApproovService.getMessageSignature(message.toString());
+                request = request.newBuilder().header(messageSigningConfig.targetHeader, signature).build();
```
Once we have the signature, call the 'generateHeaderValue' method on the MessageSigningConfig to generate the string value for the header instead of just adding the signature directly.

jexh committed:
Update ApproovMessageSigningConfig.java

Proposed changes that will need to be discussed before being adopted in a final form.

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/ApproovMessageSigningConfig.java:
```diff
@@ -0,0 +1,125 @@
+import java.util.ArrayList;
+import okhttp3.Request;
+
+/* Add the following interfaces */
```
I've added a bunch of changes here to indicate the new approach. The idea is to construct a message and a signature that can be used to correctly generate and verify the signature on the server side. To do this, while building the headers that will be added to the message we record the names of the headers and the counts of the values as a string. The generated string is then added to the message that will be signed.

The intention then is that this value is added to the header used to hold the signature along with the signature itself. The server side can then break apart the value and use it to reconstruct the message that should be used to verify the signature.

Unfortunately this implementation has a flaw, as headers not in a message but that may influence the behaviour of an endpoint can be freely added to the request and no detection would be made. Ultimately, I think the only legitimate defense against this is that the server side signature chieck also needs to know the set of headers that are expected to be included in the signature and their order. In essence, I don't think it is really possible to have the client side safely communicate to the server the set of headers that it should check in order to evaluate the legitimacy of a signature.

In summary, Ivo and Richard, I think we need some more discussions before we can fix on a nimplementation.

jexh committed:
Update ApproovMessageSigningConfig.java

New approach

jexh committed:
Update ApproovService.java

Use new message signing scheme

Commit:
Fix typos and update gradle files to conform to SDK 34

jexh committed:
Update DefaultMessageSigningConfigFactory.java

Added pseudocode for the debug capability Richard requested
Also switch to adding the sha256 of the body to the signature message instead of the body itself.

jexh commented on approov-service/src/main/java/io/approov/service/okhttp/DefaultMessageSigningConfigFactory.java:
```diff
@@ -0,0 +1,195 @@
+import java.util.ArrayList;
+import okhttp3.Request;
+
+/* Add the following interfaces */
```
This file gives a rough implementation of what it should look like - it is not proper Java as I don't have a working build env. The interfaces and classes will need their own files because they are all public.

Also, all comments need to be fleshed out and some code needs to be implemented from the comments that are provided.

Overall though, this looks like a good solution which allows debugging on the server side if a flag is set in the client and otherwise minimises the server side configuration that is required to get the signture check working. It also enables us to extend the communicated data using the same header in the near future.

Commit:
Initial changes stashed

Commit:
Rename variable

Commit:
Stash changes

johannesschneiders committed:
Approov 3.4 message signing updates from approov/combined-quickstart-service-java-okhttp.

johannesschneiders committed:
Merge branch 'main' into feature/message-signing

johannesschneiders committed:
Fix build.

johannesschneiders committed:
Update min Android SDK to 23 to match Approov SDK 3.4. Update Gradle version.

johannesschneiders committed:
Add tests from combined-quickstart-service-java-okhttp.

johannesschneiders committed:
Tidy up and POM version update.

johannesschneiders committed:
Remove TODO.

johannesschneiders committed:
Don't log the token.

johannesschneiders committed:
Update dependency versions.

johannesschneiders committed:
Comment fixes.

johannesschneiders committed:
Remove debug-logging. Tidy-up.

johannesschneiders committed:
Don't log anything that contains an Approov token.

johannesschneiders commented:
Superseded by [Feature/network intercept pinning2 #8](https://github.com/approov/approov-service-okhttp/pull/8).
